assets/tutorials/2023-07-15-gams-blog.jl

```diff
@@ -0,0 +1,366 @@
+            parsefile_dataframe(joinpath(DATA_DIR, "data_JKL.json"), (:j, :k, :l)),
+            parsefile_dataframe(joinpath(DATA_DIR, "data_KLM.json"), (:k, :l, :m)),
+        )
+    end
+end
+dataframe_data = DataFrameData(2_200);
+
+# Using the dataframe data structure, we can compactly write an equivalent JuMP
+# formulation:
+
+function dataframe_formulation(data::DataFrameData)
+    ijklm = DataFrames.innerjoin(
+        DataFrames.innerjoin(data.IJK, data.JKL; on = [:j, :k]),
+        data.KLM;
+        on = [:k, :l],
+    )
+    model = Model(Gurobi.Optimizer)
+    set_silent(model)
+    ijklm[!, :x] = @variable(model, x[1:size(ijklm, 1)] >= 0)
+    for df in DataFrames.groupby(ijklm, :i)
+        @constraint(model, sum(df.x) >= 0)
+    end
+    optimize!(model)
+    return model
+end
+
+# This formulation doesn't look like the nested summation mathematics that GAMS
+# originally formulated their model as, but it arguably just as readable,
+# particularly if the `IJKLM` columns were meaningfully related to the business
+# logic.
+
+# This is much faster, taking just over 0.1 seconds:
+
+@time dataframe_formulation(dataframe_data);
+@time dataframe_formulation(dataframe_data);
+
+# ## Scaling
+
+# Let's now compare the four different formulations over a range of `n` values:
+
+import Plots
+function timings()
+    N = [100, 200, 400, 700, 1_100, 1_600, 2_200, 2_900]
+    time_original = Float64[]
+    time_intuitive = Float64[]
+    time_fast = Float64[]
+    time_dataframe = Float64[]
+    for n in N
+        ## Original model
+        original_data = OriginalData(n)
+        start = time()
+        intuitive_formulation(original_data)
+        push!(time_original, time() - start)
+        ## Tuple models
+        tuple_data = TupleData(n)
+        start = time()
+        intuitive_formulation(tuple_data)
+        push!(time_intuitive, time() - start)
+        start = time()
+        fast_formulation(tuple_data)
+        push!(time_fast, time() - start)
+        ## DataFrame model
+        dataframe_data = DataFrameData(n)
+        start = time()
+        dataframe_formulation(dataframe_data)
+        push!(time_dataframe, time() - start)
+    end
+    return Plots.plot(
+        N,
+        [time_original time_intuitive time_fast time_dataframe];
+        labels = hcat(
+            "Intuitive (not type stable)",
+            "Intuitive (type stable)",
+            "Fast",
+            "DataFrame",
+        ),
+        xlabel = "N",
+        ylabel = "Solution time (s)",
+    )
+end
+
+# ```julia
+# timings()
+# ```
+# <img src="/assets/tutorials/gams/scaling.svg">
+
+# The dataframe formulation is asymptotically better. These results demonstrate
+# how benchmarking different modeling systems is difficult, and choosing an
+# appropriate data structure for a model is too. It's likely that Pyomo and
+# Gurobipy would similarly benefit from using `pandas` to perform the inner join
+# instead of using nested for-loop approach.
```

Comment:
Suggested replacement paragraph:

The dataframe formulation is asymptotically faster. Once we understand that the bottleneck in this benchmark is equivalent to an inner join, it is not difficult to address it, given that general-purpose languages like Julia and Python have libraries specialized for this task. Pyomo and Gurobipy would likely benefit from a similar optimization.

Reply:
"Once we understand" -> "Once one understands" ?